feat(gatsby): implement `sourceNodesStatefully` lifecycle and garbage collect stale nodes for stateless source plugins when using refresh hook (#19663)
gatsbyjs/gatsby: docs/docs/creating-a-source-plugin.md

source directory and compile the code. All plugins maintained in the Gatsby repo
follow this pattern.

Source plugins need to create Nodes for data that they would like to make available in Gatsby's GraphQL queries. To do this, you need to implement one of the following APIs in your plugin's `gatsby-node.js` file.

### `sourceNodes`

> In general, the lifecycle where your plugin should create Nodes is the [`sourceNodes`](/docs/node-apis/#sourceNodes) API. To do this, you can call the `createNode` action as illustrated in the example below.

Reviewer: ```suggestion
In general, the lifecycle where your plugin should create Nodes is the [`sourceNodes`](/docs/node-apis/#sourceNodes) API. For the majority of apps, using `sourceNodes` is the appropriate and sufficient choice. If your application is reacting in real-time to changes from an external data source, please look at the documentation for sourceNodesStatefully to determine if that is a better fit for your use case.

To do this, you can call the `createNode` action as illustrated in the example below.
```

}
```

Your plugin should create _all_ nodes in this lifecycle.

The `sourceNodes` function for your plugin _may_ be called multiple times while running `develop`. Because of this, they shouldn't register file watchers or open websockets as this will be done multiple times and could result in memory leaks.

### `sourceNodesStatefully`

> The [`sourceNodesStatefully`](/docs/node-apis/#sourceNodesStatefully) lifecycle should be used by your plugin if your data source allows subscribing to changes. In this case, your plugin is responsible for listening to these changes and calling `createNode` or `deleteNode` appropriately. Let's take a look at an example.

Reviewer: ```suggestion
The [`sourceNodesStatefully`](/docs/node-apis/#sourceNodesStatefully) lifecycle should be used by your plugin if your data source supports subscribing to changes. In this case, your plugin is responsible for listening to these changes and calling `createNode` or `deleteNode` appropriately. Let's take a look at an example.
```

```javascript:title=gatsby-node.js
exports.sourceNodesStatefully = async ({ actions }) => {
  await createInitialNodes({ actions })
  subscribeToLiveUpdates({ actions })
}

const createInitialNodes = async ({ actions }) => {
  // Get current data
  const data = await someApi.getData()

  // Process data into nodes
  data.forEach(datum => actions.createNode(processDatum(datum)))
}

const subscribeToLiveUpdates = ({ actions, getNode }) => {

>   // Setting up subscriptions. Note that this is done in separate function.

Reviewer: ```suggestion
  // Setting up subscriptions. Note that this is done in a separate function.
```

- **Add event-based sync**. Some data sources keep event logs and are able to return a list of objects modified since a given time. If you're building a source plugin, you can store
  the last time you fetched data using
  [`setPluginStatus`](/docs/actions/#setPluginStatus) and then only sync down nodes that have been modified since that time. [gatsby-source-contentful](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-source-contentful) is an example of a source plugin that does this.

> - **Proactively fetch updates**. One challenge when developing locally is that a developer might make modifications in a remote data source, like a CMS, and then want to see how it looks in the local environment. Typically they will have to restart the `gatsby develop` server or hit the [refresh endpoint](/docs/environment-variables/#reserved-environment-variables) to see changes. If data source allows subscribing to data changes events, you should implement the [`sourceNodesStatefully`](/docs/node-apis/#sourceNodesStatefully) API instead of `sourceNodes`. This will ensure that plugin won't try to create new subscriptions on data refreshes managed by Gatsby.

Reviewer: ```suggestion
- **Proactively fetch updates**. One challenge when developing locally is that a developer might make modifications in a remote data source, like a CMS, and then want to see how it looks in the local environment. Typically they will have to restart the `gatsby develop` server or hit the [refresh endpoint](/docs/environment-variables/#reserved-environment-variables) to see changes. If the data source allows you to subscribe to data change events, you should implement the [`sourceNodesStatefully`](/docs/node-apis/#sourceNodesStatefully) API instead of `sourceNodes`. This will ensure that plugin won't try to create new subscriptions on data refreshes managed by Gatsby.
```
